infra/gen-expr.rkt

```diff
@@ -183,117 +209,164 @@
+                  prec? round? allow-cond? [out-proc expr->bare])
   (define i 1)
   (for* ([c (in-range number)]
-         [expr (gen-layer ops depth exhaustive? allow-cond?)] [prec (from-list precs exhaustive?)] [rnd (from-list rnd-modes exhaustive?)])
+         [expr (gen-layer ops depth allow-cond?)] 
+         [prec (from-list precs)] 
+         [rnd (from-list rnd-modes)])
     (let-values ([(exprs* args*)  ; full expr list
-                    (let-values ([(exprs argss) (assign-terminals expr (curryr gensyms 'arg 1) exhaustive?)])
-                      (for/fold ([full-exprs '()] [arg-list '()] #:result (values full-exprs arg-list)) ; iterate through constant combinations
-                                ([expr* exprs] [args argss] #:when #t
-                                 [final (assign-known expr* (if exhaustive? consts (random-consts (unbound-in-expr expr*) consts prec exhaustive?)) exhaustive? #t)])
-                        (values (append full-exprs (list final)) (append arg-list (list args)))))])
+                    (let-values ([(exprs argss) 
+                                    (assign-terminals expr (curryr gensyms 'arg 1))])
+                      (for/fold  ; iterate through constant combinations
+                        ([full-exprs '()] [arg-list '()] #:result (values full-exprs arg-list)) 
+                        ([expr* exprs] [args argss] #:when #t 
+                          [final 
+                            (assign-known expr* 
+                              (if (exhaustive?)  
+                                 consts 
+                                 (random-consts (unbound-in-expr expr*) consts prec)) #t)])
+                        (values (append full-exprs (list final)) 
+                                (append arg-list (list args)))))])
       (for ([expr* exprs*] [args args*])
-        (let* ([prec-prop (if prec? (list (format ":precision ~a" prec)) '())]
-               [round-prop (if round? (list (format ":round ~a" (rand-from-list rnd-modes))) '())]
-               [name (if (> depth 3) (format "\"Random ~a\"" i) "")]) 
+        (let ([props (append (if prec? (list ':precision prec) '())
+                             (if round? (list ':round (rand-from-list rnd-modes)) '())
+                             (if (> depth 3) (list ':name (format "\"Random ~a\"" i)) '()))])
           (set! i (add1 i))
-          (out-proc expr* args name (append prec-prop round-prop))
+          (out-proc expr* args props)
           (gensym-count 1))))))
 
+;; Pretty formatter
+
+(define (pretty-props props)
+  (for/list ([(prop name) (in-dict (apply dict-set* '() props))])
+    (format "~a ~a" prop name)))
+
+(define (pretty-expr-helper expr) ; don't call pretty-format twice
+  (define vtor
+    (struct-copy visitor default-transform-visitor
+      [visit-! (λ (vtor props body #:ctx ctx)
+                 `(! ,@(pretty-props props) ,(visit/ctx vtor body ctx)))]))
+  (visit vtor expr))
```

Comment:
Is this pattern the same as using define/transform-expr, but without the ctx argument? Because we can definitely add a macro to make that easier.

Ideally, I'd like to structure the fpcore-visitor library so that nobody needs to actually struct-copy the visitors.

Reply:
Oops. Forgot to change this. There are already syntax rule pairs like `visit` and `visit/ctx` so maybe making similar pairs of macros would be useful.